### DISCUSSION TITLE
I am looking for examples of automation in HA with Solax value change

### DISCUSSION THREAD

**Discussion by GpMissi** [wills106/homeassistant-solax-modbus#41] [Q&A]
Hi, I would like to see some examples on how to modify the values ​​"Battery Discharge Max Current" and "Battery Minimum Capacity" through an automation of HA. Thank you

(Forgive my English, I use google translate)

**Comment by GpMissi** [comment]
Has anyone created an automation in HA to change the "Battery Discharge Max Current" and "Battery Minimum Capacity" parameters? Any idea how to do it?

**Reply by JoeAstor** [reply]
I did this. Only to put it back at 6 to 10% doesn´t work yet. Because inverter in sleep modus.
Tryed solax awake battery button, but didn´t work this morning...

alias: 'Solax: Behoud van batterij voor 6u'
description: ''
trigger:
  - platform: numeric_state
    entity_id: sensor.solax_battery_capacity
    below: '22'
condition:
  - condition: and
    conditions:
      - condition: state
        entity_id: select.solax_charger_use_mode
        state: Self Use Mode
      - condition: time
        weekday:
          - mon
          - tue
          - wed
          - thu
          - fri
        before: '05:50:00'
        after: '00:00:00'
action:
  - service: number.set_value
    data:
      value: '20'
    target:
      entity_id: number.solax_battery_minimum_capacity
  - wait_for_trigger:
      - platform: time
        at: '05:55:00'
  - service: button.press
    data: {}
    target:
      entity_id: button.solax_battery_awaken
  - delay:
      hours: 0
      minutes: 0
      seconds: 10
      milliseconds: 0
  - service: number.set_value
    data:
      value: '10'
    target:
      entity_id: number.solax_battery_minimum_capacity
mode: single

**Reply by GpMissi** [reply]
thank you :-)

**Comment by Mouxy** [comment]
I have some automations running based on the forecast for next day solar production.

Basically if the solar forecast is >15kWh I set the battery max charge to 50% from grid during the night (cheaper electricity) this way I don't need to worry about changing the modes myself

**Reply by ahb-bs** [reply]
Hi - I've an idea of what I need to do but you've stated exactly what I want to do!!  When tomorrows forecast looks like it'll reach max battery, I want to automatically dial back the max capacity.

Any chance I could get some tips or code please?

**Reply by Mouxy** [reply]
I did try to simplify as much as I could to remove loads of variables and this has been working pretty well.

I use the solar forecast (you need to pay for the api key but is not expensive) and if the forecast is above 20kWh I set by battery to charge only to 50%


alias: Set Battery Charge Limit 50%
description: Sets battery charing limit to 50% based on solar production forecast
trigger:
  - platform: time
    at: '22:30:00'
condition:
  - type: is_energy
    condition: device
    device_id: da2e1c59837f814ee46edd1cba51e944
    entity_id: sensor.energy_production_tomorrow
    domain: sensor
    above: 20
action:
  - service: number.set_value
    data:
      value: '50'
    target:
      entity_id: number.solax_forcetime_period_1_max_capacity
mode: single


About the number of writes to EEPROM, SolaX documentation says 1 million times about the number of writes and strangely that warning is missing in the X1 AC for example but my assumption is that the limit still applies.

"(...)But the EEprom has the write times limit,Usually 1Million times(...)"

**Comment by richsimpson** [comment]
I am playing with this stuff now. I think I sorted but its been a few nights of seeing what works and what the various modes onthe inverter do. I have set a slider for my desired battery SOC, Which I can easily alter and then use elsewhere in automations. 
I plan to automate the slider to know what SOC I need in the storage battery based on any of my own personal routines, and the expected solar that  will produce, plus the time of year etc.

Step 1 for all of this is the slider.
Just a simple slider from 30 to 100%. Being what I want the battery to have to start the next day. Then I later work on a query that check current SOC versus the Desired state of Charge. And sets a sensor for on or off. (I say I need 50% but the battery is only 30% = ON, or its above where I need it to be (Battery at 80% I only need 50% = OFF).

Happy to share the work and full code etc but it is still early days. Plus I am doing it in Node Red though its the same sort of automation for Home Assistant.

To be honest though, its very much a work in progress still as finding scenarios that don't work as I intend them to and so fix them.

This is what I have as the slider in Home Assistant
![image](https://user-images.githubusercontent.com/51749826/176890115-5342d093-f8d1-4e39-809a-77788ec43119.png)
Its just a simple helper 
![image](https://user-images.githubusercontent.com/51749826/176891116-ad3b93b9-63c7-4ac8-a9cf-11129bbd4cfb.png)

This is the Node Red flow that sets the sensor as on or off based on my slider
![image](https://user-images.githubusercontent.com/51749826/176889944-3cc20cc0-b4e6-4149-b779-aed1448c8c22.png)

Then I use that within a flow to set into certain modes. This is when my EV is not charging.
![image](https://user-images.githubusercontent.com/51749826/176890027-5c0fdfd8-4321-4215-9c9f-b028b3c26b7d.png)

As soon as the EV is charging, a different flow takes over to make other decisions. 
![image](https://user-images.githubusercontent.com/51749826/176890371-e4780db1-917c-4153-b9bc-ffb267b1fd25.png)

Finally the various buttons I have created to trigger manually or within the automations.
![image](https://user-images.githubusercontent.com/51749826/176890461-a88131d5-f5e0-490d-9bc5-8a20f13300f0.png)

**Reply by gittan443** [reply]
Nice one! Could you share your node-red schema?

I couldn't get the values from HA into node-red, did you do anything special to enable that? I can't see the solax modbus sensor values from within node-red.

**Reply by richsimpson** [reply]
Hi,

Happy to share. It may have developed a bit since the last post. 

I integrated the Solax into HASS witih Wills awesome integration and from there the sensors are all just available in Node Red. Home Assistant items rather than and special nodes for the integration. Sensors, input selects, services etc. Such as here:
![image](https://github.com/wills106/homeassistant-solax-modbus/assets/51749826/51db3b92-4ce1-40ec-86fc-f46fa2cad084)

I will check over the notes in current flow and make sure it all makes sense and post here. Somehow

What part are you aiming to replicate? Or just in general? I have solar stuff spread over different flows so may be able to send the important bits only to avoid adding it to your and getting loads of errors.

**Reply by richsimpson** [reply]
Here is the flow, lots of comments within the comment boxes to explain what each section is doing.
https://pastebin.com/MjavUsvT

I use a node called cronplus throughout which is an inject node that runs at staged times. You can replace with with another method to kick the processes of or install that node. Same as inject I think and subsequently I think you can make the main inject do the same thing but I never went back to look.

**Comment by PutoPunko** [comment]
Hi,

I created a couple automations, the first one reduces charge current based on main breaker current:

alias: Bateria - Límit de càrrega (set)
description: ''
trigger:
  - platform: numeric_state
    entity_id: sensor.smartlink_main_current
    above: '27'
    for:
      hours: 0
      minutes: 0
      seconds: 0
  - platform: time_pattern
    minutes: /10
condition:
  - condition: numeric_state
    entity_id: sensor.smartlink_main_current
    above: '25'
  - condition: numeric_state
    entity_id: sensor.solax_battery_power_charge
    above: '500'
  - condition: numeric_state
    entity_id: sensor.solax_grid_export
    below: '500'
action:
  - if:
      - condition: numeric_state
        entity_id: sensor.smartlink_main_current
        above: '30'
    then:
      - service: number.set_value
        data:
          value: '1'
        target:
          entity_id: number.solax_battery_charge_max_current
    else:
      - if:
          - condition: and
            conditions:
              - condition: numeric_state
                entity_id: sensor.smartlink_main_current
                above: '25'
              - condition: numeric_state
                entity_id: sensor.solax_bms_charge_max_current
                above: '7'
        then:
          - service: number.set_value
            data:
              value: '6'
            target:
              entity_id: number.solax_battery_charge_max_current
        else:
          - if:
              - condition: and
                conditions:
                  - condition: numeric_state
                    entity_id: number.solax_battery_charge_max_current
                    below: '7'
                  - condition: numeric_state
                    entity_id: sensor.smartlink_main_current
                    above: '25'
            then:
              - service: number.set_value
                data:
                  value: '1'
                target:
                  entity_id: number.solax_battery_charge_max_current
mode: single

The second one resets charge current:

alias: Bateria - Límit de càrrega (reset)
description: ''
trigger:
  - platform: numeric_state
    entity_id: sensor.smartlink_main_current
    for:
      hours: 0
      minutes: 15
      seconds: 0
    below: '18'
  - platform: numeric_state
    entity_id: sensor.smartlink_main_current
    for:
      hours: 0
      minutes: 1
      seconds: 0
    below: '6'
  - platform: time_pattern
    minutes: /5
  - platform: numeric_state
    entity_id: sensor.solax_grid_export
    for:
      hours: 0
      minutes: 5
      seconds: 0
    above: '200'
condition:
  - condition: numeric_state
    entity_id: sensor.time_automation_bateria_limit_de_carrega_reset
    above: '1800'
  - condition: numeric_state
    entity_id: sensor.time_automation_bateria_limit_de_carrega_duplica
    above: '1800'
action:
  - if:
      - condition: numeric_state
        entity_id: sensor.solax_battery_charge_max_current
        below: '6'
    then:
      - service: number.set_value
        data:
          value: '6'
        target:
          entity_id: number.solax_battery_charge_max_current
    else:
      - if:
          - condition: numeric_state
            entity_id: sensor.solax_battery_charge_max_current
            below: '10'
        then:
          - service: number.set_value
            data:
              value: '10'
            target:
              entity_id: number.solax_battery_charge_max_current
        else:
          - if:
              - condition: numeric_state
                entity_id: sensor.solax_battery_charge_max_current
                below: '13.99'
            then:
              - service: number.set_value
                data:
                  value: '14'
                target:
                  entity_id: number.solax_battery_charge_max_current
mode: single

And I have a couple extra more (I can copy them if required):

- Change Force time use minimum battery on month (summer and winter have increased energy consumption due to HVAC)
- Change discharge limit based on schedule (so on more expensive tariff more discharge is all0wed)

best regards,

**Comment by infradom** [comment]
Nice to see some nice examples here! I have not looked at them in detail, but one word of warning: most Solax parameter changes are written to EEPROM, which has a limited number of write cycles. So if you want a guaranteed lifetime of more than 10 years, we should not modify parameters more than a couple of times a day (a total of 10 updates a day should be fine however).  I am not sure if updating a parameter with exactly the same value counts as a write cycle. We could make a small modification to our integration so that updating a parameter with an identical value as the old one starts no update on the inverter.

**Reply by PutoPunko** [reply]
That's why I use a template to count last execution and have it as a condition as a safety measure. You need to use conditions with care and double check. An EEPROM warning from excesive writings would be welcome.

 - name: Time_automation.arribem_a_casa
       unit_of_measurement: "s"
       state: "{{(as_timestamp(now()) - as_timestamp(state_attr('automation.arribem_a_casa','last_triggered')))|int }}"

**Reply by richsimpson** [reply]
> Nice to see some nice examples here! I have not looked at them in detail, but one word of warning: most Solax parameter changes are written to EEPROM, which has a limited number of write cycles. So if you want a guaranteed lifetime of more than 10 years, we should not modify parameters more than a couple of times a day (a total of 10 updates a day should be fine however). I am not sure if updating a parameter with exactly the same value counts as a write cycle. We could make a small modification to our integration so that updating a parameter with an identical value as the old one starts no update on the inverter.

Yes, I spotted that in the readme file and got nervous. That is what lead to including the check on each of the changes to make sure I didn't send any instructions to go to the same current state and use one of my 100,000 lives.

**Comment by ahb-bs** [comment]
Thank you all - this has spurred me on to (carefully) use automation to change the values.

So same as stated above, if solar is great than x then set battery to y%. I've added a lot of tests to ensure that it only writes when it needs to.

One question I have is how does Rich's Force Charge button work? I can only think of changing the times on force charge but that requires constant monitoring and resetting.

Have I missed a setting/option on this?

Again, appreciate everyone's input - especially with the weather the way it is in the UK and minimising feedin!

**Reply by richsimpson** [reply]
Hi, The Force Charge Button I set up is a button created in Node Red (so in Home Assistant Too).

I could use it to set that mode but include some other scenarios. It just helps neaten my automation. And I can use more logical (to me) names.

Charge Mode 1 - Force Charge. This is checking if in manual mode. If not set it. Then check if in Force Charge (manual mode) and sets it if not. Then the house is running on grid and force charging batteries. At the point I get to the battery charge level I want (A manual slider controlled by forecasts etc eg 40% at the moment), then the inverter will then change back over into the Stop Charge and Discharge (manual mode). 
Now the inverter is going to hold the current battery level and continue to run from the grid.

Now the battery is at a decent level for my needs it will hold that through to 0030hrs then switch back to default mode and power the house from the storage batteries. 

For my use case, chances are the force charge 1 or 2 is only going to be where the EV i going to charge, so likely to be on for the 4 hours, or to top up the storage batteries, and so again likely to be on for an hour or more. Only ever starts at 2030hrs from the automations.

I can also just press my dashboard button to do the same thing but its unlikely. I did last week as battery was a bit lower than I would have liked to get through to the following dawn.

**Reply by ahb-bs** [reply]
Ah, so I'm assuming that you're hours of Force update are not set?

I'm trying to minimise the changes to the settings so have the Force charge times set as 23:30-05:30 (Octopus Intelligent Tariff).

To make it charge outside of these hours (when I the car is charging and I can add to my battery as well) I'd need to update the hours.  I was hoping for a secondary option to force charge at will, ignoring the hours but that doesn't look like the case.

At the moment I'm simply changing it to Back Up Mode outside of the force charge period - no home battery charge but any house uses the grid so makes it last longer.

**Reply by richsimpson** [reply]
I found the generic charge times were causing issues so I have the time set as nil (well 2030-2030) and then when I want to force charge the automation is calling manual mode and force charge then goes back later to self use. I only call it during my cheap time though. 

Having the times in was making my battery charge every night during the cheap slot. And then when it finished it sat there using grid instead of the battery.

**Reply by jsantosdias** [reply]
I am also having trouble with force time schedules. How did you set those to nil? Thank you!

**Comment by jsantosdias** [comment]
Well, I got it to work with this settings:
![image](https://user-images.githubusercontent.com/3714106/232445325-6f8bf2bc-9a9e-424d-86fe-422930fadece.png)

Every time I want to charge, I only change to Force Time, and get back to Self Use when I want to stop...
